src/gatsby-node.js

```diff
@@ -0,0 +1,181 @@
+/**
+ * Implement Gatsby's Node APIs in this file.
+ *
+ * See: https://www.gatsbyjs.com/docs/node-apis/
+ */
+// You can delete this file if you're not using it
+
+/**
+ * You can uncomment the following line to verify that
+ * your plugin is being loaded in your site.
+ *
+ * See: https://www.gatsbyjs.com/docs/creating-a-local-plugin/#developing-a-local-plugin-that-is-outside-your-project
+ */
+
+import { fetchStrapiContentTypes, fetchEntities, fetchEntity } from './fetch';
+import { downloadMediaFiles } from './download-media-files';
+import {
+  buildMapFromNodes,
+  buildNodesToRemoveMap,
+  getEndpoints,
+  makeParentNodeName,
+} from './helpers';
+import { createNodes } from './normalize';
+
+const LAST_FETCHED_KEY = 'timestamp';
+
+exports.onPreInit = () => console.log('Loaded gatsby-source-strapi-plugin');
+
+exports.sourceNodes = async (
+  {
+    actions,
+    createContentDigest,
+    createNodeId,
+    reporter,
+    getCache,
+    store,
+    cache,
+    getNodes,
+    getNode,
+  },
+  pluginOptions
+) => {
+  const { schemas } = await fetchStrapiContentTypes(pluginOptions);
+
+  const { deleteNode, touchNode } = actions;
+
+  const ctx = {
+    strapiConfig: pluginOptions,
+    actions,
+    schemas,
+    createContentDigest,
+    createNodeId,
+    reporter,
+    getCache,
+    getNode,
+    getNodes,
+    store,
+    cache,
+  };
+
+  const existingNodes = getNodes().filter(
+    (n) => n.internal.owner === `gatsby-source-strapi` || n.internal.type === 'File'
+  );
+
+  existingNodes.forEach((n) => touchNode(n));
+
+  const endpoints = getEndpoints(pluginOptions, schemas);
+
+  const lastFetched = await cache.get(LAST_FETCHED_KEY);
+
+  const allResults = await Promise.all(
+    endpoints.map(({ kind, ...config }) => {
+      if (kind === 'singleType') {
+        return fetchEntity(config, ctx);
+      }
+
+      return fetchEntities(config, ctx);
+    })
+  );
+
+  let newOrExistingEntries;
+
+  // Fetch only the updated data between run
+  if (lastFetched) {
+    // Add the updatedAt filter
+    const deltaEndpoints = endpoints.map((endpoint) => {
+      return {
+        ...endpoint,
+        queryParams: {
+          ...endpoint.queryParams,
+          // TODO
+          filters: {
+            updatedAt: { $gt: lastFetched },
+          },
+        },
+      };
+    });
+
+    newOrExistingEntries = await Promise.all(
+      deltaEndpoints.map(({ kind, ...config }) => {
+        if (kind === 'singleType') {
+          return fetchEntity(config, ctx);
+        }
+
+        return fetchEntities(config, ctx);
+      })
+    );
+  }
+
+  const data = newOrExistingEntries || allResults;
```

Comment:
Does this still fetch all nodes on each run? Shouldn't there be a conditional to not fetch `allResults` if there's a `lastFetched` key?

Reply:
Well we need to fetch all the results in order to know which content has been deleted.

Reply:
Does allResults mean everything all node data in Strapi? If so, needing to fetch that to determine what was deleted means the plugin doesn't really have incremental builds support.
Unless I'm misunderstanding, this code fetches all data on cold builds then on subsequent cached builds it again fetches all data and then additionally double fetches anything that changed since the first build. Is that right or am I misunderstanding?

Reply:
On subsequent cached build it uses the delta fetched data to create/update nodes. The first call is only used to know which nodes should be deleted.

So on cold build we use all the data and on then we only use the other fetches to have smaller rebuilds

Reply:
Nice 👍 good to hear that 😄 Does that happen inside fetchEntities? I'm having trouble finding it

Reply:
Not it happens with the filer `updatedAt`

Reply:
Thanks 👍 I get that part but I'm not sure how `allResults` on line 71 doesn't refetch all data on each build. On cold builds that's what pulls in all initial nodes right? I'm not following how it doesn't re-fetch all nodes again on warm builds.

Reply:
`allResults` are always fetched it is used to know which nodes we need to delete, however on subsequent build if we have already requested the API the data we use is the result of the delta fetches from L.85.

Is that an issue? Should I make an update?

If you check L.96 you'll see that we use the delta results if it exists otherwise the full results.

Reply:
Would it be interesting to add this behaviour only when `GATSBY_IS_PREVIEW` is true or should we always do it?